<a href="https://colab.research.google.com/github/INVERATE/projet_amazon/blob/main/Deep%20Learning/etape5.finit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Enregistrez le code dans le fichier « keras-test.py » dans le dossier « keras-test »
from __future__ import print_function
# Télécharger Keras
import keras
# Télécharger les dossiers de formation et de test MNIST
from keras.datasets import mnist
# Télécharger le modèle séquentiel
from keras.models import Sequential
# Télécharger les couches des cellules neuronales
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K
# Nombre de caractéristiques de données différentes : chiffres 0-9
num_classes = 10
# Nombre de laissez-passer pour la formation du réseau de neurones
epochs = 12
# Nombre de données utilisées lors d’un passage
batch_size = 128
# Dimensions des images d’entrée (28 x 28 pixels par image)
img_rows, img_cols = 28, 28
# Télécharger les formations et les tests
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
if K.image_data_format() == 'channels_first':
    train_images = train_images.reshape(train_images.shape[0], 1, img_rows, img_cols)
    test_images = test_images.reshape(test_images.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    train_images = train_images.reshape(train_images.shape[0], img_rows, img_cols, 1)
    test_images = test_images.reshape(test_images.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

# Définir le type de données en nombre décimal
train_images = train_images.astype('float32')
test_images = test_images.astype('float32')
# Normaliser les données des images
train_images /= 255
test_images /= 255
print('train_images shape:', train_images.shape)
print(train_images.shape[0], 'train samples')
print(test_images.shape[0], 'test samples')
# Convertir les vecteurs de classe en matrices de classe binaires
train_labels = keras.utils.to_categorical(train_labels, num_classes)
test_labels = keras.utils.to_categorical(test_labels, num_classes)
# Créer un modèle
model = Sequential()
# Ajouter des couches au modèle
model.add(Conv2D(2, kernel_size=(3, 3),   # couches avec fonction d’activation ReLU
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))
# Compiler le modèle
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adam(),
              metrics=['accuracy'])
# Former le modèle
model.fit(train_images, train_labels,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(test_images, test_labels))
# Évaluer le modèle
score = model.evaluate(test_images, test_labels, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
train_images shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/12


KeyboardInterrupt: 

In [ ]:
# Enregistrez le code dans le fichier « keras-test.py » dans le dossier « keras-test »
from __future__ import print_function
# Télécharger Keras
import keras
# Télécharger les dossiers de formation et de test MNIST
from keras.datasets import mnist
# Télécharger le modèle séquentiel
from keras.models import Sequential
# Télécharger les couches des cellules neuronales
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K

from scipy.sparse import load_npz
import pandas as pd

import numpy as np

X_train_npz = np.load('X_train_tfidf.npz')
X_train = pd.DataFrame.from_dict({item: [X_train_npz[item]] for item in X_train_npz.files}, orient='index')
X_test_npz = np.load('X_test_tfidf.npz')
X_test = pd.DataFrame.from_dict({item: [X_test_npz[item]] for item in X_test_npz.files}, orient='index')

#X_train = load_npz("X_train_tfidf.npz")
Y_train = pd.read_csv("y_train.csv")
#X_test = load_npz("X_test_tfidf.npz")
Y_test = pd.read_csv("y_test.csv")

model = Sequential()
model.add(Dense(12, input_shape=(8,), activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

from tensorflow.keras.utils import plot_model

plot_model(model, to_file="model.png", show_shapes=True, show_layer_names=False, show_layer_activations=True)

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

history = model.fit(X_train, Y_train, validation_split=0.2, epochs=50, batch_size=10)

from matplotlib import pyplot as plt

plt.plot(history.history['accuracy'], color='#066b8b')
plt.plot(history.history['val_accuracy'], color='#b39200')
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'val'], loc='upper left')
plt.show()

predictions = model.predict(X_test)

predictions[0]

predictions = (model.predict(X_test) > 0.5).astype(int)

for i in range(5):
	print('%s => Prédit : %d,  Attendu : %d' % (X_test.iloc[i].tolist(), predictions[i], Y_test.iloc[i]))

_, accuracy = model.evaluate(X_test, Y_test)
print('Accuracy: %.2f' % (accuracy*100))

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


ValueError: Invalid dtype: object

In [ ]:
nn_architecture = [
    {"input_dim": 2, "output_dim": 4, "activation": "relu"},
    {"input_dim": 4, "output_dim": 6, "activation": "relu"},
    {"input_dim": 6, "output_dim": 6, "activation": "relu"},
    {"input_dim": 6, "output_dim": 4, "activation": "relu"},
    {"input_dim": 4, "output_dim": 1, "activation": "sigmoid"},
]

def init_layers(nn_architecture, seed = 99):
    np.random.seed(seed)
    number_of_layers = len(nn_architecture)
    params_values = {}

    for idx, layer in enumerate(nn_architecture):
        layer_idx = idx + 1
        layer_input_size = layer["input_dim"]
        layer_output_size = layer["output_dim"]

        params_values['W' + str(layer_idx)] = np.random.randn(
            layer_output_size, layer_input_size) * 0.1
        params_values['b' + str(layer_idx)] = np.random.randn(
            layer_output_size, 1) * 0.1


def sigmoid(Z):
    return 1/(1+np.exp(-Z))

def relu(Z):
    return np.maximum(0,Z)

def sigmoid_backward(dA, Z):
    sig = sigmoid(Z)
    return dA * sig * (1 - sig)

def relu_backward(dA, Z):
    dZ = np.array(dA, copy = True)
    dZ[Z <= 0] = 0;
    return dZ;



def single_layer_forward_propagation(A_prev, W_curr, b_curr, activation="relu"):
    Z_curr = np.dot(W_curr, A_prev) + b_curr

    if activation is "relu":
        activation_func = relu
    elif activation is "sigmoid":
        activation_func = sigmoid
    else:
        raise Exception('Non-supported activation function')

    return activation_func(Z_curr), Z_curr



def full_forward_propagation(X, params_values, nn_architecture):
    memory = {}
    A_curr = X

    for idx, layer in enumerate(nn_architecture):
        layer_idx = idx + 1
        A_prev = A_curr

        activ_function_curr = layer["activation"]
        W_curr = params_values["W" + str(layer_idx)]
        b_curr = params_values["b" + str(layer_idx)]
        A_curr, Z_curr = single_layer_forward_propagation(A_prev, W_curr, b_curr, activ_function_curr)

        memory["A" + str(idx)] = A_prev
        memory["Z" + str(layer_idx)] = Z_curr

    return A_curr, memory


def get_cost_value(Y_hat, Y):
    m = Y_hat.shape[1]
    cost = -1 / m * (np.dot(Y, np.log(Y_hat).T) + np.dot(1 - Y, np.log(1 - Y_hat).T))
    return np.squeeze(cost)

def get_accuracy_value(Y_hat, Y):
    Y_hat_ = convert_prob_into_class(Y_hat)
    return (Y_hat_ == Y).all(axis=0).mean()




def single_layer_backward_propagation(dA_curr, W_curr, b_curr, Z_curr, A_prev, activation="relu"):
    m = A_prev.shape[1]

    if activation is "relu":
        backward_activation_func = relu_backward
    elif activation is "sigmoid":
        backward_activation_func = sigmoid_backward
    else:
        raise Exception('Non-supported activation function')

    dZ_curr = backward_activation_func(dA_curr, Z_curr)
    dW_curr = np.dot(dZ_curr, A_prev.T) / m
    db_curr = np.sum(dZ_curr, axis=1, keepdims=True) / m
    dA_prev = np.dot(W_curr.T, dZ_curr)

    return dA_prev, dW_curr, db_curr



def full_backward_propagation(Y_hat, Y, memory, params_values, nn_architecture):
    grads_values = {}
    m = Y.shape[1]
    Y = Y.reshape(Y_hat.shape)

    dA_prev = - (np.divide(Y, Y_hat) - np.divide(1 - Y, 1 - Y_hat));

    for layer_idx_prev, layer in reversed(list(enumerate(nn_architecture))):
        layer_idx_curr = layer_idx_prev + 1
        activ_function_curr = layer["activation"]

        dA_curr = dA_prev

        A_prev = memory["A" + str(layer_idx_prev)]
        Z_curr = memory["Z" + str(layer_idx_curr)]
        W_curr = params_values["W" + str(layer_idx_curr)]
        b_curr = params_values["b" + str(layer_idx_curr)]

        dA_prev, dW_curr, db_curr = single_layer_backward_propagation(
            dA_curr, W_curr, b_curr, Z_curr, A_prev, activ_function_curr)

        grads_values["dW" + str(layer_idx_curr)] = dW_curr
        grads_values["db" + str(layer_idx_curr)] = db_curr

    return grads_values


def update(params_values, grads_values, nn_architecture, learning_rate):
    for layer_idx, layer in enumerate(nn_architecture):
        params_values["W" + str(layer_idx)] -= learning_rate * grads_values["dW" + str(layer_idx)]
        params_values["b" + str(layer_idx)] -= learning_rate * grads_values["db" + str(layer_idx)]

    return params_values;


def train(X, Y, nn_architecture, epochs, learning_rate):
    params_values = init_layers(nn_architecture, 2)
    cost_history = []
    accuracy_history = []

    for i in range(epochs):
        Y_hat, cashe = full_forward_propagation(X, params_values, nn_architecture)
        cost = get_cost_value(Y_hat, Y)
        cost_history.append(cost)
        accuracy = get_accuracy_value(Y_hat, Y)
        accuracy_history.append(accuracy)

        grads_values = full_backward_propagation(Y_hat, Y, cashe, params_values, nn_architecture)
        params_values = update(params_values, grads_values, nn_architecture, learning_rate)

    return params_values, cost_history, accuracy_history

<>:45: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:47: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:90: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:92: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:45: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:47: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:90: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:92: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-3-85d6de635c12>:45: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if activation is "relu":
<ipython-input-3-85d6de635c12>:47: SyntaxWarning: "is" with a literal. Did you mean "=="?
  elif activation is "sigmoid":
<ipython-input-3-85d6de635c12>:90: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if activation is "relu":
<ipython-input-3-85d6de635c12>:92: SyntaxWarning: "is" with a literal. Did you mean "=="?
  elif activation is "sigmoid":


In [ ]:
# first neural network with keras tutorial
from numpy import loadtxt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from scipy.sparse import load_npz
import pandas as pd

X_train = load_npz("X_train_tfidf.npz")
Y_train = pd.read_csv("y_train.csv")
X_test = load_npz("X_test_tfidf.npz")
Y_test = pd.read_csv("y_test.csv")

# define the keras model
model = Sequential()
model.add(Dense(12, input_shape=(8,), activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))


# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])



# fit the keras model on the dataset
model.fit(X_train, Y_train, epochs=150, batch_size=10)


# evaluate the keras model
_, accuracy = model.evaluate(X_train, Y_train)
print('Accuracy: %.2f' % (accuracy*100))


AxisError: axis 1 is out of bounds for array of dimension 1

In [3]:
from scipy.sparse import load_npz
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import pandas as pd


X_train = load_npz("X_train_tfidf.npz").toarray()
Y_train = pd.read_csv("y_train.csv")['y_train']
X_test = load_npz("X_test_tfidf.npz").toarray()
Y_test = pd.read_csv("y_test.csv")['y_test']

# Vérification de la dimension d'entrée
input_dim = X_train.shape[1]  # nombre de colonnes = nombre de features TF-IDF


model = Sequential()
model.add(Dense(64, input_shape=(input_dim,), activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))


model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


model.fit(X_train, Y_train, epochs=10, batch_size=10)


_, accuracy = model.evaluate(X_test, Y_test)
print('Accuracy: %.2f' % (accuracy * 100))


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.0979 - loss: -1.6218
Epoch 2/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.1067 - loss: -31.2996
Epoch 3/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.1082 - loss: -139.0155
Epoch 4/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.0906 - loss: -409.6908
Epoch 5/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.1148 - loss: -873.7919
Epoch 6/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.0969 - loss: -1663.0658
Epoch 7/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 1s 9ms/step - accuracy: 0.1078 - loss: -2782.0251
Epoch 8/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 1s 9ms/step - accuracy: 0.1064 - loss: -4263.7354
Epoch 9/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.1077 - loss: -6345.6577
Epoch 10/10
80/80 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.0910 - loss: -8826.6924
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.0662 - loss: -11598.4746  
Accuracy: 6.97


# 2 couches de 64 puis 32 neurones


In [10]:
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import pandas as pd
from scipy.sparse import load_npz
import numpy as np


# Encodage
X_train = load_npz("X_train_tfidf.npz").toarray()
Y_train = pd.read_csv("y_train.csv")['y_train']
X_test = load_npz("X_test_tfidf.npz").toarray()
Y_test = pd.read_csv("y_test.csv")['y_test']
Y_train_cat = to_categorical(Y_train - 1, num_classes=5)
Y_test_cat  = to_categorical(Y_test  - 1, num_classes=5)

input_dim = X_train.shape[1]

model = Sequential([
    Dense(64, input_shape=(input_dim,), activation='relu'),
    Dense(32, activation='relu'),
    Dense(5, activation='softmax')
])

model.compile(
    loss='categorical_crossentropy',
    optimizer='adam',
    metrics=['categorical_accuracy']
)

model.fit(X_train, Y_train_cat, epochs=10, batch_size=32,
          validation_split=0.1)

_, accuracy = model.evaluate(X_test, Y_test_cat)
print(f'Multi-class accuracy: {accuracy:.2%}')

Epoch 1/10


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


23/23 ━━━━━━━━━━━━━━━━━━━━ 2s 20ms/step - categorical_accuracy: 0.6075 - loss: 1.5505 - val_categorical_accuracy: 0.6625 - val_loss: 1.3147
Epoch 2/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - categorical_accuracy: 0.6359 - loss: 1.2370 - val_categorical_accuracy: 0.6625 - val_loss: 1.0920
Epoch 3/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - categorical_accuracy: 0.6669 - loss: 0.9977 - val_categorical_accuracy: 0.6625 - val_loss: 1.0341
Epoch 4/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - categorical_accuracy: 0.6526 - loss: 0.8901 - val_categorical_accuracy: 0.6625 - val_loss: 1.0011
Epoch 5/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - categorical_accuracy: 0.6216 - loss: 0.7931 - val_categorical_accuracy: 0.6625 - val_loss: 0.9767
Epoch 6/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - categorical_accuracy: 0.6499 - loss: 0.6535 - val_categorical_accuracy: 0.6750 - val_loss: 0.9837
Epoch 7/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 1s 13ms/step - categorical_accuracy: 0.8345 - loss: 0.5004 - val_categ

# Optimisation du modèle

In [34]:
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import pandas as pd
from scipy.sparse import load_npz
import numpy as np
from tensorflow.keras.optimizers import Adam, RMSprop, Adamax, Nadam, Lion


# Encodage
X_train = load_npz("X_train_tfidf.npz").toarray()
Y_train = pd.read_csv("y_train.csv")['y_train']
X_test = load_npz("X_test_tfidf.npz").toarray()
Y_test = pd.read_csv("y_test.csv")['y_test']
Y_train_cat = to_categorical(Y_train - 1, num_classes=5)
Y_test_cat  = to_categorical(Y_test  - 1, num_classes=5)

input_dim = X_train.shape[1]

model = Sequential([
    Dense(128, input_shape=(input_dim,), activation='relu'),
    Dense(64, activation='relu'),
    Dense(5, activation='softmax')
])

model.compile(
    loss='categorical_crossentropy',
    #optimizer=Adam(learning_rate=0.00099),
    optimizer=Nadam(learning_rate=0.0009),
    metrics=['categorical_accuracy']
)

model.fit(X_train, Y_train_cat, epochs=10, batch_size=32,
          validation_split=0.1)

_, accuracy = model.evaluate(X_test, Y_test_cat)
print(f'Multi-class accuracy: {accuracy:.2%}')

Epoch 1/10


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


23/23 ━━━━━━━━━━━━━━━━━━━━ 3s 37ms/step - categorical_accuracy: 0.5620 - loss: 1.5356 - val_categorical_accuracy: 0.6625 - val_loss: 1.1595
Epoch 2/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 1s 16ms/step - categorical_accuracy: 0.6170 - loss: 1.1431 - val_categorical_accuracy: 0.6625 - val_loss: 1.0296
Epoch 3/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - categorical_accuracy: 0.6032 - loss: 1.0273 - val_categorical_accuracy: 0.6625 - val_loss: 0.9944
Epoch 4/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 1s 15ms/step - categorical_accuracy: 0.6558 - loss: 0.7445 - val_categorical_accuracy: 0.6750 - val_loss: 0.9735
Epoch 5/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 1s 14ms/step - categorical_accuracy: 0.7287 - loss: 0.6015 - val_categorical_accuracy: 0.6875 - val_loss: 0.9964
Epoch 6/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 1s 18ms/step - categorical_accuracy: 0.8915 - loss: 0.3790 - val_categorical_accuracy: 0.7125 - val_loss: 1.0066
Epoch 7/10
23/23 ━━━━━━━━━━━━━━━━━━━━ 1s 17ms/step - categorical_accuracy: 0.9801 - loss: 0.2236 - val_ca

En ajoutant des couches (5 avec 128 --> 12) l'accuracy diminue (64%).
En augmentant à 0.1 le learning rate l'accuracy diminue diminue aussi (51%) , avec O.5, 0.99, 0.00099 l'accuracy diminue moins (66%). 0.02 ---> 62%.

Diminution du learning rate ! augmentation de l'accuarcy.
0.00099-->68.66%
0.0009-->69.14%

La modification du learning rate ne change pas grand chose.

Optimizer RMSprop, Nadam même résultat
Nadam 0.0009 ---> 69.15% !!
Adamax, Lion moins bien (66%, 62%)